### ISSUE TITLE
Allow arguments to be passed into JavaScript scripts

### ISSUE THREAD

**Issue by ombre42** [robotframework/SeleniumLibrary#323]
Selenium allows arguments to be passed along with JavaScript that can be referenced in the script using the injected List variable named `arguments`. _Execute Javascript_, _Execute Async Javascript_, and _Wait For Condition_ do not support this feature. This would be useful for passing WebElements and data structures into scripts

When we implement _Get Webelement(s)_, people will naturally want to pass the WebElements as argument(s) to a script. Currently, one can use _Assign Id To Element_ to make a WebElement easy to reference if JS if it is difficult to reference is JS.

Implementing this without breaking test cases or adding keywords would require using keyword arguments for _Execute Javascript_ and _Execute Async Javascript_. Because they have variable arguments, we cannot add a positional argument. Keyword arguments require RF 2.8 to work.

There is a workaround for this missing functionality on [StackOverflow](http://stackoverflow.com/questions/21256845/is-there-a-way-to-provide-arguments-to-execute-javascript-in-robotframework/21273554#21273554)

**Comment by pekkaklarck** [comment]
Requiring RF 2.8 shouldn't be a problem because it's not that new release anymore. I don't think minimum RF requirement needs to be raised, though, but obviously that needs to be documented in kw doc.

**Comment by ombre42** [comment]
After getting far into adding this feature using keyword arguments, I realized there is a serious backwards incompatibility issue. Assignments or comparisons in JavaScript are triggering Robot Framework's free keyword argument processing. Example:
`Execute Javascript    var elem = document.getElementById('foo'); elem.selected = true;`
Fails with message _'var elem ' is not a valid keyword argument_. The equals sign would have to be escaped:
`Execute Javascript    var elem \= document.getElementById('foo'); elem.selected \= true;`
I pushed [a commit](https://github.com/ombre42/robotframework-selenium2library/commit/3dac91b5e3160d5e1866d7cad7ac7cd3c4688c20) if you want to see the progress thus far. I am not going to do anything more until we have a consensus.

I see three options to implement this feature:
1) Implement with keyword arguments, breaking _many_ scripts and always needing to escape '='s
2) Remove the capability to split the code in multiple cells (change sig to `execute_javascript(code, *args)`). This will break tests that split the code (far less than option 1), unfortunately with the vague error _WebDriverException: Message: u'JavaScript error'_.
3) Add two keywords _Execute Javascript With Args_ and _Execute Async Javascript With Args_ that have the arguments (code, *args), thus not supporting breaking the script into multiple cells. This would not break any test but certainly adds to keyword clutter.

@pekkaklarck @emanlove @j1z0  I want hear your opinions and anyone else's on this. Maybe we should post to the user group and have a poll like Ed mentioned?

**Comment by ombre42** [comment]
I just thought of one thing that would make option 2 less painful. If arguments were passed and _code_ does not contain the string _arguments_ then fail with message _Arguments were passed and not referenced. Splitting the script into multiple cells is no longer supported_

**Comment by pekkaklarck** [comment]
Comments to analysis by @ombre42:

1) The motivation to allow splitting the script to multiple arguments in old SeleniumLibrary was to ease executing longer scripts. This was especially important in old HTML test data days when you couldn't easily have a long string even in one row due to columns. Even today splitting multiple lines of JavaScript like below can be handy:

``` robotframework
*** Test Cases ***
Example
    Execute Javascript
    ...    var foo = 'example';
    ...    var result = doSomething(foo);
    ...    if (result == 42)
    ...        // and so on
```

2) Passing arguments to JavaScript sounds like a nice feature. Because longer JavaScript code can be easily constructed using `Catenate` keyword, losing the above ability to pass the code in pieces shouldn't be that big a problem in general.

3) It would be good to avoid backwards incompatible changes is possible. @ombre42 already presented an idea how to detect a situation where user has passed the code in pieces. Instead of failing, it would be better to emit a deprecation warning and behave like earlier, though. Alternatively we could deprecate passing the code in multiple pieces first and add this functionality only in the next release.

4) The requirement to escape `=` makes using kwargs pretty much a no-go in this particular use case. My trivial example above has four equal signs already.

5) Would a good signature be:

``` python
def execute_javascript(code, arguments):
    # ...
```

or

``` python
def execute_javascript(code, *arguments):
    # ...
```

The latter would have the same problem as the current signature if the keyword ever needs more arguments. It would be a little more convenient than creating a list first, though, but if it's fine to require the code to be given as one and not in pieces, that shouldn't matter too much either.

From backwards compatibility point of view, the former signature would cause problem if the code is passed in as more than two pieces. It would thus need to be something like:

``` python
def execute_javascript(code, arguments, *deprecated):
    # ...
```

**Comment by ombre42** [comment]
Thank you so much for your input Pekka. I never understood why the code-split feature. We use RIDE exclusively where I work and any formatting like you mentioned will be re-written by RIDE.
Maybe I'm short-sighted, but I do not foresee adding arguments to this in the future, so I'm leaning towards this solution, which can go into 1.6:
Change to `def execute_javascript(code, *arguments):`

```
If len(arguments) > 0
   If code contains the string 'arguments'
       process arguments as arguments to the script
   else
       process arguments as extra code to catenate with code and script will have no arguments
       log a warning
else
    run script with have no arguments
```

I think the probability of a script containing the string 'arguments' but not intending to reference the injected arguments from Selenium is quite low, but the logic must be described in the keyword doc.

We could also change it to
`def execute_javascript(code, *arguments_or_code)`
And continue support the ability to split code indefinitely based on wheter _code_ contains the string _arguments_ (and not log warnings if it doesn't). This is a little awkward, but worth mentioning.

**Comment by pekkaklarck** [comment]
Hmm, `execute_javascript(code, *arguments_or_code)` is definitely awkward but also a pretty clever idea. Not losing the ability to split long code and thus also avoiding backwards-incompatible changes (except if code contains `arguments`) would be a big win. I don't see why this wouldn't work if it is clearly documented with examples.

**Comment by ombre42** [comment]
@emanlove - Thoughts? The reason behind *arguments_or_code was that there does not seem to be much gained by removing the backwards compatibility code. That, and warnings should be used with caution. Warning about not supporting functionality in a future version is not good unless you have a plan on when and why.

**Comment by emanlove** [comment]
@ombre42: I'm still reading through this and thinking about different solutions. I tend to think having arguments will be useful and a functionality that we would want to add. There will be an issue with using your idea in that the string 'arguments' is what the javascript uses within the script to 'call' the argument. For example see [some of my AngularJS additions](https://github.com/emanlove/robotframework-selenium2library/compare/angularjs#diff-b91175814577caa11b43b4be06319f57R180). It is a great idea by the way but we would need a different separator string.

**Comment by emanlove** [comment]
Looking at the examples above I noticed the Javascript code is not delineated. For some reason, possibly because my recent use of Python webdriver, I was expecting quotes around the script to be executed. Why not use this as the delimator or separator between script and arguments. If the first argument begins and ends with a quotation mark then this marks the script and the remaining arguments are arguments for the javascript.  So for example the second execute javascript call in the following RF test does not create an alert box; it is simply an unassigned javascript string.

```
Execute Javascript With Quotes
    [Documentation]  Testing out exec Javascript with Arguments
    Execute Javascript  alert('Hello World!')
    Alert Should Be Present  text=Hello World!
    Execute Javascript  "alert('Hello World!')"
    Run Keyword And Expect Error  *  Alert Should Be Present  text=Hello World!
```

Here one could do something like ...

```
Execute Javascript With Arguments
    [Documentation]  Testing out exec Javascript with Arguments
    Execute Javascript  alert('Hello World!')
    Alert Should Be Present  text=Hello World!
    Execute Javascript  "alert('arguments[0]');"  ${argument1}  ${argument2}
    Alert Should Be Present  text=${argument1}
```

This might be an alternative to your use of the 'arguments' string, Kevin. We could even get fancy and count quotes doing something like

```
Execute MultiLine Javascript With Quotes
    [Documentation]  Testing out exec Javascript with Arguments
    Execute Javascript  alert('Hello World!')
    Alert Should Be Present  text=Hello World!
    Execute Javascript  "alert('Hello World!');
    ...    var foo = 'example';
    ...    var result = doSomething(foo);
    ...    if (result == arguments[1])
    ...        // and so on    ..."
    ...    argument1
    ...    argument2
```

**Comment by emanlove** [comment]
In trying to see if an unassigned string literal would conflict in anyway with the Javascript specification I did come across the [_"use strict";_ directive](http://ecma-international.org/ecma-262/5.1/#sec-10.1.1) and more generic concept of [Directive Prologues](http://ecma-international.org/ecma-262/5.1/#sec-14.1). Noting in the specifications for [Directive Prologues](http://ecma-international.org/ecma-262/5.1/#sec-14.1) that the semi-colon is not required explicitly it would be possible then for someone to write a test case like

```
Execute MultiLine Javascript With Use Strict
    [Documentation]  Testing out exec Javascript with Arguments
    Execute Javascript  "use strict"
    ...    var foo = 'example';
    ...    var result = doSomething(foo);
    ...    if (result == arguments[1])
    ...        // and so on    ...
```

and not intend to use the _Execute Javascript_ keyword as I propose. But I think this a minor case that can be handle by 1) checking for common directive prologues (although I can't seem to find a list of common ones nor do I think they are explicitly defined; i.e. anyone create create custom ones) and 2) users can add a semi-colon to the end thus avoiding our assumption that this is a code block and the remaining arguments are javascript arguments. We could also provide some sort of warning.

**Comment by pekkaklarck** [comment]
After thinking about possible solutions a little more, simply implementing new `Execute Javascript With Arguments` and `Execute Async Javascript With Arguments` start to sound pretty tempting.

Pros:
- No backwards-compatibility issues whatsoever.
- Clean and easy to document API.

Cons:
- Two more keywords.
- New keyword names are pretty long.

**Comment by ombre42** [comment]
@emanlove Thanks for pointing out that the "arguments" trick won't work. I was not thinking of async JS where it would always be in the script to call the callback.
If we are going to add keywords, I think "args" is an acceptable abbreviation to make the keyword name a _little_ shorter - its OK for PyCharm's spell-check and its used in the DatabaseLibrary.
I still think deprecating splitting the script for one version and changing to (code, *arguments) or (code, *args) is still on the table.
If we are going to add keyword(s), we could do exucute_javascript_with_args(code, async='False', *args) but that essentially makes async a mandatory argument and I do not like that.
I am glad both of you are so concerned about finding the ~~right~~ best solution as everything we have come up with so far has been a comprimise.

**Comment by just-be-dev** [comment]
I'd like to revive this conversation as it seems like some people may need it. 

I like some of the ideas @ombre42 had. I think I have an idea that would kinda fit in the best of both worlds but it's still a little hacky. 

I'm imaging something like this

``` python
def execute_javascript(args, code*):
  if len(code) == 0:
    code.append(args)
  elif isstr(args):
    code.insert(0, args)
  elif isinstance(args, list):
    browser.execute_script(code, args*)
  else:
    browser.execute_script(code, args)
```

(Look over my really rough psudo-code. I think you get the idea.)

So, it is a little hacky. If the argument is a string then it must be added to a list before it can be passed through. Regular objects can be passed in like normal though (like web elements). Multiple arguments have to be passed through an array. If there's no args then it's called exactly how it is today. 

What do you think @pekkaklarck, @emanlove?

**Comment by emanlove** [comment]
Have not yet had a chance to review your idea, will do so sometime in the next few days, but I agree that this change should be reconsidered.

**Comment by just-be-dev** [comment]
It was a pretty easy implementation so I went ahead and made it happen. 

When you get a chance look it over and tell me what you think.

**Comment by pekkaklarck** [comment]
I agree passing arguments to the code can be useful. Passing args before the code feels a bit odd, though. It's unfortunate this keyword accepts varargs so having a clean API without breaking backwards compatibility isn't possible. One possibility would be using kwargs like:

``` python
def execute_javascript(self, *code, **config):
    args = config.pop('args', None)
    if config:
        raise SuitableException("Message explaining only 'args' allowed in 'config'.")
    ...
```

Unfortunately using kwargs has a drawback that all normal equal signs in arguments need to be escaped with a backslash. Do you have any idea how common they are in code executed with this keyword?

**Comment by just-be-dev** [comment]
Any assumptions I made on have often equal signs were present would just be conjecture. Likewise with the multi-line support. I was just trying to fit something within the scope of what would work with the old implementation without implementing new keywords. 

Also, I actually think it's not bad to pass the args first. 

Consider this user keyword (which is cross posted from #288):

``` robotframework
*** Keywords ***
Highlight Element
    [Arguments]    ${locator}
    ${element}=    Get WebElement    ${locator}
    Execute Javascript   ${element}
    ...   element = arguments[0];
    ...   original_style = element.getAttribute('style');
    ...   element.setAttribute('style', original_style + "; background: yellow;");
    ...   setTimeout(function(){
    ...        element.setAttribute('style', original_style);
    ...   }, 300);
```

Having a large JavaScript code body would make it cumbersome to have the arguments at the end of the code segment. Putting them first actually makes it feel somewhat like writing a function. The only reason this implemention isn't ideal is that to have multiple arguments you need to have them in a list object. You can check out the implementation logic in #387.

**Comment by pekkaklarck** [comment]
Good points. Perhaps it would just be easiest to implement new `Execute Javascript with Arguments` keyword with signature `code, *args`.

If having two highly related keywords is considered a problem, using `Execute Javascript` so that code is split to multiple arguments could even be deprecated. This would take quite many releases, though:
1. Give a deprecation warning if `len(code) > 1` in 1.7.
2. Change signature to `def execute_javascript(code)` in 1.8.
3. Change signature to `def execute_javascript(code, *args)` in 1.9.
4. Deprecate `Execute Javascript with Arguments` in 1.9 or later.
5. Finally remove `Execute Javascript with Arguments` in the next release after deprecating it.

**Comment by pekkaklarck** [comment]
If we consider splitting code to multiple arguments a useful feature (it works fine in [above example](https://github.com/rtomac/robotframework-selenium2library/issues/323#issuecomment-99900171)) but still would like to allow passing arguments to the code separately in the same call, we could possibly allow that by having a special marker like `ARGUMENTS` telling where arguments start:

``` robotframework
*** Keywords ***
Highlight Element
    [Arguments]    ${locator}
    ${element}=    Get WebElement    ${locator}
    Execute Javascript
    ...   element = arguments[0];
    ...   original_style = element.getAttribute('style');
    ...   element.setAttribute('style', original_style + "; background: yellow;");
    ...   setTimeout(function(){
    ...        element.setAttribute('style', original_style);
    ...   }, 300);
    ...   ARGUMENTS
    ...   ${element}
```

This ought to be pretty backwards compatible because it's very unlikely that anyone would have a literal string like `ARGUMENTS` in an existing keyword call as a separate argument. This can be considered a hack, but at least the usage would be pretty simple. Robot also already uses similar special arguments `ELSE`, `ELSE IF` and `AND` with `Run Keyword If` and `Run Keywords`.

**Comment by just-be-dev** [comment]
You know what would be really nice?

``` robotframework
*** Keywords ***
Highlight Element
    [Arguments]    ${locator}
    ${element}=    Get WebElement    ${locator}
    Execute Javascript
    ...   original_style = ${element}.getAttribute('style');
    ...   ${element}.setAttribute('style', original_style + "; background: yellow;");
    ...   setTimeout(function(){
    ...        ${element}.setAttribute('style', original_style);
    ...   }, 300);
```

What if we did that? The implementation would be more tricky, but it's so nice and it solves all our problems without weird argument placements or killing backwards compatibility. 

Also, @pekkaklarck, I think you forgot your example, haha.

**Comment by just-be-dev** [comment]
In hindsight, that actually comes with its own set of problems. Besides the complexity, it doesn't work for passing arguments to JavaScript files. 

I don't want to take away the ability to execute multi-line javascript code inline. I'm biased here though, because I use that quite often. I guess technically the tester could use `Catenate` to join the lines of JavaScript together and pass that to `Execute Javascript`... hmm....

**Comment by pekkaklarck** [comment]
If passing the code as multiple arguments is that useful, it's definitely not a good idea to remove that functionality. Using `Catenate` would be as annoying as using `Create List` to create arguments.

Yes, I accidentally submitted my previous comment before it was finished. Edited now.

**Comment by just-be-dev** [comment]
@pekkaklarck, What if we just always passed in a map of the currently defined variables?

Going back to that old example:

``` robotframework
*** Keywords ***
Highlight Element
    [Arguments]    ${locator}
    ${element}=    Get WebElement    ${locator}
    Execute Javascript
    ...   var v = arguments[0];
    ...   original_style = v['${element}'].getAttribute('style');
    ...   v['${element}'].setAttribute('style', original_style + "; background: yellow;");
    ...   setTimeout(function(){
    ...        v.['${element}'].setAttribute('style', original_style);
    ...   }, 300);
```

Though I'll admit that I don't yet know what the performance ramifications this could have. I'll have to research if there's a heavy cost for passing in a potentially large object. 

I'll probably just post on the user-group tomorrow and get some feedback from them.

**Comment by pekkaklarck** [comment]
How would `v['${element}']` be interpreted in this example?

Asking feedback from users is definitely a good idea.

**Comment by just-be-dev** [comment]
It would be the web element returned from Get WebElement. I expect that `argument[0]` would look something like

``` javascript
{ "${element}": [Object object], ... }
```

**Comment by pekkaklarck** [comment]
I don't understand how this would actually work. Is `${element}` interpreted as a normal Robot variable or somehow specially in this context?

**Comment by pekkaklarck** [comment]
Ooooops, who put "Comment and close" button next to "Comment" button??? Reopening.....

**Comment by just-be-dev** [comment]
Well, the theory was to use built in to get a dictionary of the currently defined keywords. I could then pass that to `execute_script` as an argument which would (in theory) convert that dict to a javascript obj. Then, on the client side all the variables would be available by keys. 

I played with it a little last night, but I haven't got that implementation working yet. I'm not even sure it can properly handle a python dict. I'm working on it though, and I'll submit another PR when I'm done.

**Comment by yahman72** [comment]
Is there any updates to this? I.e. Plans to support passing "args" to java-script?
The #387 has a solution but it's not backwards compatible. 

I'd probably go for a new KW `Execute Javascript with Arguments`

**Comment by aaltat** [comment]
There are a several ways to do it. 
1) Create new keyword which supports required functionality
2) Deprecate the current keyword in release 2.0 and change the functionality in release 3.0 to support arguments
3) Deprecate the current keyword in release 2.0  and create new keywords to support arguments. In release 3.0 change the current keyword to support arguments and deprecate the new keywords. In release 4.0 remove new keywords, created in release 2.0.

There are also some other variations, but they need more work. The option 2 would be best way to go at least from maintenance point of view.

Generally thinking, the feature is a useful one, but for biwic reason it's not very fast one to make.

**Comment by arnewohletz** [comment]
Seems this enhancement didn't make it release 3.0. Are there any news?
So far, I keep adding this piece of code to each new library installation, which does it for me:
```python
@keyword
def execute_javascript_with_arguments(self, code, *arguments):
    js = self._get_javascript_to_execute(code)
    self.info("Executing JavaScript \n%s with arguments \n%s" % (
    js, str(*arguments)))
    return self.driver.execute_script(js, *arguments)
```

**Comment by aaltat** [comment]
You are correct and this is not yet implemented. Would you be willing to provide a PR?

**Comment by arnewohletz** [comment]
Sure, I can do that. Thanks for your appreciation!

**Comment by aaltat** [comment]
By a quick look, I thought we had a design in the issue, but after some more reading we don't actually have a design. At least I don't know what should I expect in the OR. Before implementing the PR, we should have a rough idea about the design. And before that I should read how JavaScript arguments actually work.

**Comment by aaltat** [comment]
We did agree the design in the #1147 PR. In short it is this: 
1. Support current style:
```
Execute JavaScript
...    element = arguments[0];
...    original_style = element.getAttribute('style');
...    element.setAttribute('style', original_style + "; background: yellow;");
...    setTimeout(function(){
...       element.setAttribute('style', original_style);
...    }, 300);
```
2. Support arguments in the end:
```
Execute Javascript
...    element = arguments[0];
...    original_style = element.getAttribute('style');
...    element.setAttribute('style', original_style + "; background: yellow;");
...    setTimeout(function(){
...       element.setAttribute('style', original_style);
...    }, 300);
...    ARGUMENTS
...    ${element}    ${unused_arg}
```
3. Support changing the order of arguments and code:
```
Execute Javascript
...    ARGUMENTS
...    ${element}    ${unused_arg}
...    JAVASCRIPT
...    element = arguments[0];
...    original_style = element.getAttribute('style');
...    element.setAttribute('style', original_style + "; background: yellow;");
...    setTimeout(function(){
...       element.setAttribute('style', original_style);
...    }, 300);
```
4. Support defining the JAVASCRIPT at start (optional)
```
Execute Javascript
...    JAVASCRIPT
...    element = arguments[0];
...    original_style = element.getAttribute('style');
...    element.setAttribute('style', original_style + "; background: yellow;");
...    setTimeout(function(){
...       element.setAttribute('style', original_style);
...    }, 300);
...    ARGUMENTS
...    ${element}    ${unused_arg}
```
For more details see the discussion in #1147

**Comment by pekkaklarck** [comment]
Looks good. What's the use case for `JAVASCRIPT` marker?

**Comment by aaltat** [comment]
Option number 3, when markers are swapped.

**Comment by pekkaklarck** [comment]
Well, what's use case for that? =)

**Comment by aaltat** [comment]
I had a discussion and sometimes it's feels more natural to define the arguments first in JavaScript. For me it feels, when looking from Python side, totally unnatural, but then again I am not that familiar with JavaScript.

**Comment by pekkaklarck** [comment]
I found that odd order as well. If someone finds it useful, and is willing to implement it, I don't see much problems with it, though. Only somewhat serious drawback is making the documentation of the keyword more complicated.

**Comment by aaltat** [comment]
There has been users that have found it useful and it could be that @horsewithnoname1985 provides a PR. 

I do agree that explaining this in the documentation is somewhat challenging. It might be worth while to consider to add a section to the library documentation and not only keyword documentation.

**Comment by aaltat** [comment]
Fixed in #1176
